File: Data Structures & Algorithms/search-for-word/submission-1.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])



        def dfs(i,r,c):
            if i >= len(word):
                return True
            if r >= rows or r < 0 or c >= cols or c < 0 or word[i] != board[r][c]:
                return False
            else:
                board[r][c] = '#'
                res = (dfs(i+1, r+1, c) 
                    or dfs(i+1, r, c+1) 
                    or dfs(i+1, r-1, c) 
                    or dfs(i+1, r, c-1))
                board[r][c] = word[i]
                return res

        for r in range(rows):
            for c in range(cols):
                if dfs(0,r,c):
                    return True

        return False

File: Data Structures & Algorithms/max-area-of-island/submission-1.cpp
class Solution {
   public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int ROWS = grid.size();
        int COLS = grid[0].size();
        int maxArea = 0;
        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == 1) {
                    maxArea = max(maxArea, dfs(grid, r, c));
                }
            }
        }
        return maxArea;
    }

   private:
    int dfs(vector<vector<int>>& grid, int r, int c) {
        int ROWS = grid.size();
        int COLS = grid[0].size();
        if (r < 0 || c < 0 || r >= ROWS || c >= COLS || grid[r][c] == 0) {
            return 0;
        }
        grid[r][c] = 0;
        return 1 + dfs(grid, r + 1, c) + dfs(grid, r - 1, c) + dfs(grid, r, c + 1) +
               dfs(grid, r, c - 1);
    }
};
